lib/Sema/CSSolver.cpp

```diff
@@ -1301,8 +1311,30 @@ static bool tryTypeVariableBindings(
       if (!typeVar->getImpl().canBindToLValue())
         type = type->getRValueType();
 
-      // Remove parentheses. They're insignificant here.
-      type = type->getWithoutParens();
+      bool removeParentheses = true;
+      // Don't remove parentheses if this is a function argument
+      // in Swift >= 4 mode, because after SE-0110 parentheses play
+      // a role in function argument matching, e.g.
+      //
+      // T0 bound to (((Int, Int)) -> Int) -> [Int];
+      // T0 conv (T5) -> Int
+      //
+      // If (after SE-0110) we remove parentheses from binding:
+      // T5 bond to (((Int, Int)) -> Int)
+
+      if (!cs.getASTContext().isSwiftVersion3()) {
```

Comment:
Even in Swift 3 mode, we should be able to perform this logic, since matchTypes() is more permissive, no?